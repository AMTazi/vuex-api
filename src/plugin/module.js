import pluginActions from './actions'
import vue from 'vue'
import axios from 'axios'
const state = {}

const actions = {
  [pluginActions.request]: ({ commit }, { url, method, baseURL, ...options }) => {
    commit(pluginActions.request, options)
    axios({ method: method || 'GET', baseURL: baseURL || '', url }).then(resp => {
      commit(pluginActions.success, { ...options, resp: resp })
    }).catch(err => {
      commit(pluginActions.error, { ...options, err: err.response })
    })
  },
  [pluginActions.clear]: ({ commit }, keyPath) => {
    commit(pluginActions.clear, keyPath)
  }
}

const mutations = {
  [pluginActions.request]: (state, { keyPath }) => {
    const obj = { ...state[keyPath], state: 'loading' }
    vue.set(state, keyPath, obj)
  },
  [pluginActions.success]: (state, { keyPath, resp }) => {
    const obj = { ...state[keyPath], state: 'success', firstCallDone: true, resp }
    vue.set(state, keyPath, obj)
  },
  [pluginActions.error]: (state, { keyPath, err }) => {
    const obj = { ...state[keyPath], state: 'error', err }
    vue.set(state, keyPath, obj)
  },
  [pluginActions.clear]: (state, keyPath) => {
    const obj = {}
    vue.set(state, keyPath, obj)
  }
}

export const module = {
  state,
  actions,
  mutations
}

export default module
