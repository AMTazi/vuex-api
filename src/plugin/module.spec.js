/**
 * Created by vouill on 10/24/17.
 */
// import Vue from 'vue/dist/vue.js'
import module from './module'
import pluginActions from './actions'

const testRequest = { url: 'testUrl', method: 'testMethod', baseURL: 'baseUrl' }

describe('Module Actions', () => {
  it('should process request', () => {
    const commit = jest.fn()
    module.actions[pluginActions.request]({ commit }, testRequest)
    expect(commit.mock.calls).toMatchSnapshot()
  })

  it('should process clear', () => {
    const commit = jest.fn()
    module.actions[pluginActions.clear]({ commit }, 'someKeypath')
    expect(commit.mock.calls).toMatchSnapshot()
  })
})

describe('Module mutations', () => {
  let state = {}
  const keyPath = 'someKeyPath'
  beforeEach(() => {
    state = {}
  })
  it('should process request', () => {
    module.mutations[pluginActions.request](state, { keyPath })
    expect(state).toMatchSnapshot()
  })

  it('should process success', () => {
    module.mutations[pluginActions.success](state, { keyPath, resp: { date: 'fakeData' } })
    expect(state).toMatchSnapshot()
  })

  it('should process error', () => {
    module.mutations[pluginActions.error](state, { keyPath, err: { data: 'fakeError' } })
    expect(state).toMatchSnapshot()
  })

  it('should process clear', () => {
    module.mutations[pluginActions.clear](state, { keyPath })
    expect(state).toMatchSnapshot()
  })

  it('should process request -> success', () => {
    module.mutations[pluginActions.clear](state, { keyPath })
    module.mutations[pluginActions.success](state, { keyPath, resp: { date: 'fakeData' } })
    expect(state).toMatchSnapshot()
  })

  it('should process request -> error', () => {
    module.mutations[pluginActions.clear](state, { keyPath })
    expect(state).toMatchSnapshot()
    module.mutations[pluginActions.error](state, { keyPath, err: { data: 'fakeError' } })
    expect(state).toMatchSnapshot()
  })
})
