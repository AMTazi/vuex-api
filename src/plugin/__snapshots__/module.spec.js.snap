// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Module Actions should process clear 1`] = `
Array [
  Array [
    "Clear_Api_Call",
    "someKeypath",
  ],
]
`;

exports[`Module Actions should process request 1`] = `
Array [
  Array [
    "Request_Api_Call",
    Object {},
  ],
]
`;

exports[`Module mutations should process clear 1`] = `
Object {
  "[object Object]": Object {},
}
`;

exports[`Module mutations should process error 1`] = `
Object {
  "someKeyPath": Object {
    "err": Object {
      "data": "fakeError",
    },
    "state": "error",
  },
}
`;

exports[`Module mutations should process request -> error 1`] = `
Object {
  "[object Object]": Object {},
}
`;

exports[`Module mutations should process request -> error 2`] = `
Object {
  "[object Object]": Object {},
  "someKeyPath": Object {
    "err": Object {
      "data": "fakeError",
    },
    "state": "error",
  },
}
`;

exports[`Module mutations should process request -> success 1`] = `
Object {
  "[object Object]": Object {},
  "someKeyPath": Object {
    "firstCallDone": true,
    "resp": Object {
      "date": "fakeData",
    },
    "state": "success",
  },
}
`;

exports[`Module mutations should process request 1`] = `
Object {
  "someKeyPath": Object {
    "state": "loading",
  },
}
`;

exports[`Module mutations should process success 1`] = `
Object {
  "someKeyPath": Object {
    "firstCallDone": true,
    "resp": Object {
      "date": "fakeData",
    },
    "state": "success",
  },
}
`;
